File: chap10_cellregulation/lac_operon.py

#   -------------------------------------------------------------------
# 
#   Plot solutions of the lac operon model 
# 
#    For Chapter 10, Section 10.2.5 of
#    Keener and Sneyd, Mathematical Physiology, 3rd Edition, Springer.
# 
#    Written by James Keener and James Sneyd
# 
#   -------------------------------------------------------------------

import numpy as np
import matplotlib.pyplot as plt
from scipy.integrate import solve_ivp

# Initial conditions and time span
init = [0.01, 0.01, 0.01, 0.0, 0.0]
tspan = np.linspace(0, 2000, 1000)

# Parameters for ODE solver
tol = 1e-9

# Function to calculate Le
def getLe(t):
    if t < 1000:
        return 0.04 * t / (50 + t)
    else:
        return 0.04 * 1000 / (50 + 1000) * 1 / (1 + (t - 1000) / 500)

# Function for the system of ODEs
def rhs(t, x):
    M, B, P, A, L = x
    
    aA, bA, gamA = 1.76e4, 2.15e4, 0.52
    aB, gamB = 1.66e-2, 2.26e-2
    aP, gamP = 10.0, 0.65
    aM, gamM = 9.97e-4, 0.41
    aL, gamL = 2880.0, 2.26e-2
    Kaa, K, KA, KL, KLe, KL1 = 2.52e4, 6000.0, 1.95, 9.7e-7, 0.26, 1.81
    
    Le = getLe(t)
    
    dM = aM * (1.0 + Kaa * A**2) / (K + Kaa * A**2) - gamM * M
    dB = aB * M - gamB * B
    dP = aP * M - gamP * P
    dA = aA * B * L / (KL + L) - bA * B * A / (KA + A) - gamA * A
    dL = aL * P * Le / (KLe + Le) - aA * B * L / (KL + L) - gamL * L
    
    return [dM, dB, dP, dA, dL]

# Solve the system of ODEs
sol = solve_ivp(rhs, [tspan[0], tspan[-1]], init, method='LSODA', t_eval=tspan, atol=tol, rtol=tol, max_step=0.01)

# Calculate Le for each time point
Le = np.array([getLe(t) for t in sol.t])

# Plot the results
plt.figure(1)
plt.plot(sol.t, sol.y[3], label='A', linewidth=2)
plt.ylabel('A')
plt.xlabel('Time')

plt.twinx()
plt.plot(sol.t, Le, label='Le', linewidth=2, color='orange')
plt.ylabel('Le')
plt.legend()

File: chap10_cellregulation/lac_operon.m

%  -------------------------------------------------------------------
%
%  Plot solutions of the lac operon model 
%
%   For Chapter 10, Section 10.2.5 of
%   Keener and Sneyd, Mathematical Physiology, 3rd Edition, Springer.
%
%   Written by James Keener and James Sneyd
%
%  -------------------------------------------------------------------

close all
clear all
clc

set(0,                           ...
    'defaultaxesfontsize', 20,   ...
    'defaultaxeslinewidth', 2.0, ...
    'defaultlinelinewidth', 2.0);
global tmax
tmax = 10000;
nT = 1001;
init=[0.01 0.01 0.01 0.0 0];
tspan = linspace(0,tmax,nT );
 
 opts = odeset('AbsTol',1e-9,'RelTol',1e-9,'MaxStep',0.01); % be careful with the numerics
[T,Y] = ode15s(@(t,x)rhs(t,x),tspan,init,opts);

for i=1:nT 
    Le(i) = getLe(T(i));  % super inefficient, but it's fast enough and it works
end

figure(1)
    plot(T,Y(:,4),'LineWidth',2)
    ylabel('A (mM)')
    yyaxis right
    plot(T,Le,'LineWidth',2)
ylabel('L_e (mM)')
box off

figure(2)
semilogy(Le,Y(:,4))
xlabel('L_e')
ylabel('A')
box off


 %%
function out = rhs(t,x)
M = x(1);
B = x(2);
P = x(3);
A = x(4);
L = x(5);

aA=1.76e4; bA=2.15e4; gamA=0.52;
aB=1.66e-2; gamB=2.26e-2;
aP=10.0; gamP=0.65;
aM=9.97e-4; gamM=0.41;
aL=2880.0; gamL=2.26e-2;
Kaa=2.52e4; K=6000.0; KA=1.95; KL=9.7e-7; KLe=0.26; KL1=1.81;

Le = getLe(t);

out(1)=aM*(1.0+Kaa*(A)^2)/(K+Kaa*(A)^2) - (gamM)*M;
out(2)=aB*M-(gamB)*B;
out(3)=aP*M-(gamP)*P;
out(4)=aA*B*L/(KL+L)-bA*B*A/(KA+A)-(gamA)*A;
out(5)=aL*P*(Le)/(KLe+Le) - aA*B*L/(KL+L) - (gamL)*L;

out = out';
end

%% 
function Le = getLe(t)
global tmax
tmax

if t<=tmax/2
    Le = 0.04*t./(0.1*tmax+t);
else
    Le = 0.02*tmax/(0.1*tmax+tmax/2)./(1+(t-tmax/2)/(tmax/4));
    %Le=1./(1+(t-tmax/2)/(tmax/4));
end
end